data/code/lib.c

```diff
@@ -589,3 +593,715 @@ void check_lap_ex()
+    speed *= 10.0;
+
+    z_direction_vector[0] *= speed;
+    z_direction_vector[2] *= speed;
+    add_absolute_position_vector(
+        this, z_direction_vector[0], z_direction_vector[1], z_direction_vector[2]);
+}
+
+// Resets last recorded XZ momentum before bounce liftoff;
+void reset_last_momentum(int kart_num)
+{
+    float* const last_momentum = &s_last_momenta[kart_num];
+    *last_momentum = 0;
+}
+
+void call_do_spd_ctrl_normal(char* const strat)
+{
+    KartStrat__DoSpeedCrl(strat);  // Original instruction;
+}
+
+void call_do_spd_ctrl_mod(char* const this,
+                          char* const ctrl,
+                          char* const race_manager,
+                          int kart_num)
+{
+    handle_boosts(this);
+    handle_x_adjustment(this, ctrl, race_manager, kart_num);
+    handle_y_adjustment(this, ctrl, kart_num);
+    clamp_movement_vector_descent(this);
+}
+
+void call_do_spd_ctrl(char* const this,
+                      char* const strat,
+                      char* const ctrl,
+                      char* const race_manager,
+                      int kart_num,
+                      int kart_bounce_flag)
+{
+    if (kart_bounce_flag == false)
+    {
+        call_do_spd_ctrl_normal(strat);
+    }
+    else
+    {
+        call_do_spd_ctrl_mod(this, ctrl, race_manager, kart_num);
+
+        // ASM section done here instead of its own function (has differing results?).
+        // Forcefully set registers to result of original instruction for peace of mind.
+
+        asm("lis %r4, 0x8143");
+        asm("ori %r4, %r4, 0x3a50");
+        asm("lis %r5, 0x0000");
+        asm("ori %r5, %r5, 0x0040");
+        asm("lis %r6, 0x0000");
+        asm("ori %r6, %r6, 0x0040");
+        asm("lis %r7, 0x0000");
+        asm("ori %r7, %r7, 0x0040");
+        asm("lis %r8, 0x0000");
+        asm("ori %r8, %r8, 0x0040");
+        asm("lis %r9, 0x0000");
+        asm("ori %r9, %r9, 0x000a");
+        asm("li %r10, 0x0");
+    }
+}
+
+// Currently is a two byte structure. If new materials need flags, this can be added to and extended.
+void set_kart_extended_terrain_flag(char* flag, unsigned int hash, int value)
+{
+    if (value == 0)
+    {
+        *flag = *flag & ~hash;
+    }
+    else
+    {
+        *flag = *flag | hash;
+    }
+}
+
+void set_kart_bounce_liftoff_flag(char* flag, int value)
+{
+    set_kart_extended_terrain_flag(flag, 0x2, value);
+}
+
+void set_kart_bounce_flag(char* flag, int value)
+{
+    set_kart_extended_terrain_flag(flag, 0x1, value);
+}
+
+void set_kart_bounce_flag_both(char* flag, int value)
+{
+    set_kart_bounce_liftoff_flag(flag, value);
+    set_kart_bounce_flag(flag, value);
+}
+
+int get_kart_extended_terrain_flag(char* flag, unsigned int hash)
+{
+    int terrain_flag = *flag & hash;
+    return terrain_flag;
+}
+
+int get_kart_bounce_liftoff_flag(char* flag)
+{
+    int ret = 0;
+    if (get_kart_extended_terrain_flag(flag, 0x2) != 0)
+    {
+        ret = 1;
+    }
+    return ret;
```

Comment:
It looks like the implementation could be a one liner:

```cpp
return get_kart_extended_terrain_flag(flag, 0x2) != 0;
```

Reply:
I was taught, "write code for readability first and optimize later". I suppose what one considers readability is highly subjective; I prefer my implementation, but you do not.

Truthfully, I rarely consider how the compiler is going to interpret my code, but in this case I decided that maybe this was simply a change to save precious space, so I decided to compare the compiler output - however, the compiler was already smart enough to optimize the function; both implementations are _fundamentally the same_.

Were this not the case, I'd instantly choose the faster implementation without a comment. With this in mind, I'm choosing to refactor it to your preference for consistency's sake alone (it is _your_ extender) - I'm just not entirely happy about it myself.

Reply:
It's really quite opinionated. I do value readability too, but in this case having the extra variable means having 6 lines instead of one, for no real benefit.

I just get a bit nervous when I see non-const variables. If a variable cannot be const, I always question whether it can be dropped.

What makes the longer implementation easier to read for you? Is it that a literal `1` and `0` are present? If so, it could be:

```cpp
return get_kart_extended_terrain_flag(flag, 0x2) ? 1 : 0;
```

Or:


```cpp
return get_kart_extended_terrain_flag(flag, 0x2) ? true : false;
```

Reply:
It's hard to describe, but it's to do with the verticality of it. I am not dyslexic and yet I struggle to read that; it's one thing to not understand what something the first time, and another to have to repeatedly parse something that others seem to read easily.

As a result, changing the conditional to a different, more horizontal syntax is even more challenging. 

I'll adapt to initial one-line implementation - I shouldn't complain so much about something so minor.